internal/zinc-classpath/src/main/scala/sbt/internal/inc/ScalaInstance.scala

```diff
@@ -66,6 +66,11 @@ final class ScalaInstance(
     }
   }
 
+  /** ClassLoader with just scala-library.jar. */
+  lazy val libraryLoader: ClassLoader = {
+    classpath.ClasspathUtilities.toLibraryLoader(this)
+  }
+
```

Comment:
I think it would be preferable to do this way in in `ScalaProvider`, ie change the way it creates `loader`, and then just accept `libraryLoader` and `fullLoader` here.

Reply:
Your change will end up with `scala-library` loaded separates in `loader` and `libraryLoader`. That's a functional problem if you ever need interop between programs running in those loaders. It is also a small performance problem in space and time for the VM.

Reply:
You could sort of cheat without changing the interface of `ScalaProvider` and the intervening call chain.

The launcer could create the layered classloaders as I described, but still only pass the full one into `ScalaInstance`. Here, you could look at `loader.getParent` to get the `libraryLoader` (perhaps verifying that `isIsntanceOf[LauncherLibraryClassloader]` to know that the secret handshake is complete

Reply:
If the secret handshake wasn't made (ie, we are being called by an old launcher), you could either do as you've done here, or just let the bug persist.

Reply:
That's in launcher land. I would rather not make ppl reinstall launchers. (Assuming bunch of ppl are using official sbt package, not sbt-extra)

Reply:
> The launcer could create the layered classloaders as I described, but still only pass the full one into ScalaInstance.
  
I think this is the way to go -- being able to recycle JVM optimizations that have already happened for the Scala library is important, and classloading the library again goes against that.
  
I have the impression that, sooner or later, we'll need to deal with a better way to organise classloaders for both performance and correctness. I would prefer that to happen than to work around the limitations of the current design.

Reply:
> That's a functional problem if you ever need interop between programs running in those loaders.

I need to think more on this. My intention is to use this stuff for `run` and `test`, maybe `console`. The all-included loader would be used during `compile`.

Reply:
Here's what I've got in mind: https://github.com/retronym/launcher/tree/topic/library-loader

Reply:
We could deliver both fixes: the launcher fix (that makes this part of the code trivial), and the existing fix you've got here (to fix the bug when using older launchers)

Reply:
You'd need to deal with the `xsbti.ExtendedScalaProvider` interface reflectively in SBT to avoid a `LinkageError` with older launchers.

Reply:
@jvican 
> I think this is the way to go -- being able to recycle JVM optimizations that have already happened for the Scala library is important, and classloading the library again goes against that.

I agree. I think I am now starting to see the light of what Jason is saying.

Reply:
https://github.com/sbt/launcher/pull/47